File: src/game/entity/models/Enemy.java
package game.entity.models;


import game.Bomberman;
import game.entity.Player;
import game.entity.bomb.AbstractExplosion;
import game.entity.bomb.ConfettiExplosion;
import game.events.KilledEnemyEvent;
import game.events.ScoreGameEvent;
import game.ui.panels.game.PitchPanel;

import java.util.*;

public abstract class Enemy extends Character{
    public Enemy() {
        this(null);
        setCoords(Coordinates.randomCoordinatesFromPlayer(getSize()));
    }

    public Enemy(Coordinates coordinates) {
        super(coordinates);
    }

    @Override
    protected void doInteract(Entity e) {
        if (e instanceof BomberEntity) {
            attack(e);
        }
    }

    @Override
    protected void onSpawn() {
        super.onSpawn();
        Bomberman.getMatch().increaseEnemiesAlive();
        Bomberman.getMatch().getGameTickerObservable().register(this);
    }

    @Override
    protected void onEliminated() {
        super.onEliminated();
        new KilledEnemyEvent().invoke(this);
        new ScoreGameEvent().invoke(this.getMaxHp());
    }

    @Override
    protected void onDespawn() {
        super.onDespawn();
        Bomberman.getMatch().decreaseEnemiesAlive();
        Bomberman.getMatch().getGameTickerObservable().unregister(this);
    }

    @Override
    public Set<Class<? extends Entity>> getInteractionsEntities(){
        return new HashSet<>(Collections.singletonList(Player.class));
    }

    public final void update(Object gameState) {
        super.update(gameState);

        if((boolean) gameState)
            doUpdate((Boolean) gameState);
    }

    public abstract void doUpdate(boolean gameState);

    public int getSize() {
        return PitchPanel.GRID_SIZE;
    }

    @Override
    protected Set<Class<? extends Entity>> getBasePassiveInteractionEntities() {
        return new HashSet<>(Collections.singletonList(AbstractExplosion.class));
    }

}

File: src/game/entity/models/Enemy.kt
package game.entity.models

import game.Bomberman
import game.entity.player.Player
import game.entity.bomb.AbstractExplosion
import game.entity.player.BomberEntity
import game.events.game.EnemyDespawnedGameEvent
import game.ui.panels.game.PitchPanel

abstract class Enemy(coordinates: Coordinates?) : Character(coordinates) {
    constructor() : this(null) {
        if (Bomberman.getMatch().player != null) {
            coords = Coordinates.randomCoordinatesFromPlayer(size)
        }
    }

    override fun doInteract(e: Entity?) {
        (e as? BomberEntity)?.let { attack(it) }
    }

    override fun onSpawn() {
        super.onSpawn()
        Bomberman.getMatch().increaseEnemiesAlive()
        Bomberman.getMatch().gameTickerObservable?.register(this)
    }

    override fun onDespawn() {
        super.onDespawn()
        val match = Bomberman.getMatch() ?: return
        match.decreaseEnemiesAlive()
        (match.gameTickerObservable ?: return).unregister(this)
        EnemyDespawnedGameEvent().invoke(null)
    }

    override fun getInteractionsEntities(): Set<Class<out Entity>> {
        return HashSet<Class<out Entity>>(listOf(Player::class.java))
    }

    override fun update(arg: Any?) {
        super.update(arg)
        if (arg as Boolean) doUpdate(arg)
    }

    abstract fun doUpdate(gameState: Boolean)

    override val size: Int
        get() = PitchPanel.GRID_SIZE

    override val basePassiveInteractionEntities: MutableSet<Class<out Entity>>
        get() = hashSetOf(AbstractExplosion::class.java)
}